nibabel/elemcont.py

```diff
@@ -0,0 +1,188 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+'''Containers for storing "elements" which have both a core data value as well
+as some additional meta data. When indexing into these containers it is this
+core value that is returned, which allows for much cleaner and more readable
+access to nested structures.
+
+Each object stored in these containers must have an attribute `value` which
+provides the core data value for the element. To get the element object itself
+the `get_elem` method must be used.
+'''
+from collections import MutableMapping, MutableSequence
+
+from .externals import OrderedDict
+from .externals.six import iteritems
+
+
+class Elem(object):
+    '''Basic element type has a `value` and a `meta` attribute.'''
+    def __init__(self, value, meta=None):
+        self.value = value
+        self.meta = {} if meta is None else meta
+
+
+class InvalidElemError(Exception):
+    '''Raised when trying to add an object without a `value` attribute to an
+    `ElemDict`.'''
+    def __init__(self, invalid_val):
+        self.invalid_val = invalid_val
+        message = ("Provided value '%s' of type %s does not have a 'value' "
+                   "attribute" % (self.invalid_val, type(invalid_val)))
+        super(InvalidElemError, self).__init__(message)
+
+
+class ElemDict(MutableMapping):
+    '''Ordered dict-like where each value is an "element", which is defined as
+    any object which has a `value` attribute.
+
+    When looking up an item in the dict, it is this `value` attribute that
+    is returned. To get the element itself use the `get_elem` method.
+    '''
+
+    def __init__(self, *args, **kwargs):
+        if len(args) > 1:
+            raise TypeError("At most one arg expected, got %d" % len(args))
+        self._elems = OrderedDict()
+        if len(args) == 1:
+            arg = args[0]
+            if hasattr(arg, 'get_elem'):
+                it = ((k, arg.get_elem(k)) for k in arg)
+            elif hasattr(arg, 'items'):
+                it = iteritems(arg)
+            else:
+                it = arg
+            for key, val in it:
+                self[key] = val
+        for key, val in iteritems(kwargs):
+            self[key] = val
+
+    def __getitem__(self, key):
+        return self._elems[key].value
+
+    def __setitem__(self, key, val):
+        if not hasattr(val, 'value'):
+            raise InvalidElemError(val)
+        self._elems[key] = val
+
+    def __delitem__(self, key):
```

Comment:
I believe that subclassing MutableMapping is preferred for more "custom" dict-like classes, rather than just extending the API of a standard `dict`. In particular, other standard `dict` methods won't call your custom `__getitem__` and `__setitem__` methods.